libraries/lib-cloud-audiocom/UploadService.cpp

```diff
@@ -0,0 +1,475 @@
+#include <rapidjson/writer.h>
+
+#include "OAuthService.h"
+#include "ServiceConfig.h"
+
+#include "NetworkManager.h"
+#include "Request.h"
+#include "IResponse.h"
+#include "MultipartData.h"
+
+#include "CodeConversions.h"
+
+namespace cloud::audiocom
+{
+namespace
+{
+std::string_view DeduceMimeType(const wxString& ext)
+{
+   if (ext == "wv")
+      return "audio/x-wavpack";
+   else if (ext == "flac")
+      return "audio/x-flac";
+   else if (ext == "mp3")
+      return "audio/mpeg";
+   else
+      return "audio/x-wav";
+}
+
+std::string GetUploadRequestPayload(const wxString& filePath, const wxString& projectName)
+{
+   rapidjson::Document document;
+   document.SetObject();
+
+   const wxFileName fileName(filePath);
+   const auto mimeType = DeduceMimeType(fileName.GetExt());
+
+   document.AddMember(
+      "mime",
+      rapidjson::Value(
+         mimeType.data(), mimeType.length(), document.GetAllocator()),
+      document.GetAllocator());
+
+   const auto name = audacity::ToUTF8(projectName.empty() ? fileName.GetFullName() : projectName);
+
+   document.AddMember(
+      "name",
+      rapidjson::Value(name.data(), name.length(), document.GetAllocator()),
+      document.GetAllocator());
+
+   document.AddMember(
+      "size",
+      rapidjson::Value(static_cast<int64_t>(fileName.GetSize().GetValue())),
+      document.GetAllocator());
+
+   rapidjson::StringBuffer buffer;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+   document.Accept(writer);
+
+   return std::string(buffer.GetString());
+}
+
+std::string GetProgressPayload(uint64_t current, uint64_t total)
+{
+   rapidjson::Document document;
+   document.SetObject();
+
+   document.AddMember(
+      "progress", rapidjson::Value(current / static_cast<double>(total) * 100.0),
+      document.GetAllocator());
+
+   rapidjson::StringBuffer buffer;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+   document.Accept(writer);
+
+   return std::string(buffer.GetString());
+}
+
+
+// This class will capture itself inside the request handlers
+// by a strong reference. This way we ensure that it outlives all
+// the outstanding requests.
+struct UploadOperation final :
+    UploadPromise,
+    std::enable_shared_from_this<UploadOperation>
+{
+   UploadOperation(
+      const ServiceConfig& serviceConfig, wxString fileName,
+      wxString projectName,
+      UploadService::CompletedCallback completedCallback,
+      UploadService::ProgressCallback progressCallback)
+       : mServiceConfig(serviceConfig)
+       , mFileName(std::move(fileName))
+       , mProjectName(std::move(projectName))
+       , mCompletedCallback(std::move(completedCallback))
+       , mProgressCallback(std::move(progressCallback))
+   {
+   }
+
+   const ServiceConfig& mServiceConfig;
+   
+   wxString mFileName;
+   wxString mProjectName;
+
+   UploadService::CompletedCallback mCompletedCallback;
+   UploadService::ProgressCallback mProgressCallback;
```

Comment:
Above four members can be `const`

Reply:
Only the first two. I want to be able to reset the handlers, so the associated resources can be released as well